pkg/controllers/proxyconfig/proxyconfig_controller.go

```diff
@@ -0,0 +1,200 @@
+// to connect to endpoint directly and via proxy and comparing the results with expectations.
+func checkProxyConfig(ctx context.Context, endpoint, noProxy string, clientWithProxy, clientWithoutProxy *http.Client) error {
+	endpointURL, err := url.Parse(endpoint)
+	if err != nil {
+		return err
+	}
+
+	reachableWithProxy := isEndpointReachable(ctx, endpoint, clientWithProxy)
+	reachableWithoutProxy := isEndpointReachable(ctx, endpoint, clientWithoutProxy)
+	noProxyMatchesEndpoint := strings.Contains(noProxy, endpointURL.Hostname())
+
+	if noProxyMatchesEndpoint && !reachableWithoutProxy {
+		if reachableWithProxy {
+			return fmt.Errorf("endpoint(%q) found in NO_PROXY(%q) can only be reached with proxy", endpoint, noProxy)
+		}
+		return fmt.Errorf("endpoint(%q) found in NO_PROXY(%q) is unreachable with/without proxy", endpoint, noProxy)
+	}
+
+	if !noProxyMatchesEndpoint && !reachableWithProxy {
+		if reachableWithoutProxy {
+			return fmt.Errorf("endpoint(%q) missing in NO_PROXY(%q) can only be reached without proxy", endpoint, noProxy)
+		}
+		return fmt.Errorf("endpoint(%q) is unreachable with/without proxy", endpoint)
+	}
+
+	return nil
+}
+
+// createHTTPClients returns two http clients, one with proxy and another without proxy
+func (p *proxyConfigChecker) createHTTPClients() (*http.Client, *http.Client, error) {
+	caPool, err := p.getCACerts()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs: caPool,
+	}
+
+	return &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConfig,
+				Proxy:           proxyFunc,
+			},
+		}, &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConfig,
+			},
+		}, nil
+}
+
+// getCACerts retrieves the CA bundle in openshift cluster
+func (p *proxyConfigChecker) getCACerts() (*x509.CertPool, error) {
+	caConfigmap, err := p.configMapLister.ConfigMaps(p.caConfigmapNamespace).Get(p.caConfigmapName)
+	if err != nil {
+		return nil, err
+	}
+
+	caPool := x509.NewCertPool()
+	if ok := caPool.AppendCertsFromPEM([]byte(caConfigmap.Data["ca-bundle.crt"])); !ok {
+		return nil, fmt.Errorf("unable to append system trust ca bundle")
+	}
+
+	return caPool, nil
+}
+
+// isEndpointReachable returns true if the given endpoint can be reached using the given client
+func isEndpointReachable(ctx context.Context, endpointURL string, client *http.Client) bool {
+	req, err := http.NewRequest(http.MethodGet, endpointURL, nil)
+	if err != nil {
+		klog.V(4).Infof("failed to create new http request: %v", err)
+		return false
+	}
+	reqCtx, cancel := context.WithTimeout(ctx, 5*time.Second) // avoid waiting forever
+	defer cancel()
+	req.WithContext(reqCtx)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		klog.V(4).Infof("failed to make request to endpoint: %v", err)
+		return false
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		klog.V(4).Infof("%q returned %q", endpointURL, resp.Status)
+		return false
+	}
```

Comment:
consider keeping the errors to go into the status so that you've got more information right away without having to look into the logs